### DISCUSSION TITLE
CP maximum allowable slip rate

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#28534] [Q&A Modules: Solid mechanics]
Hello, I am running a compression test on a cylindre with 300 grains, a mesh with 111,000 DoF, a height of 72 mm. I want to reach 2% strain I have set the forced displacement at 10^-5 * t and a time step to reach a maximum of 72 (starting at 1 with the help of timestepper) but I keep getting warnings of the form of : 'Maximum allowed slip increment exceeded'. I've set the slip increment to 10. Is there a way to calibrate this slip increment value to optimize the time step and loading rate and still not have to wait forever to reach at least 1% strain. Thanks !!

```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
  [file]
    type = FileMeshGenerator
    file = n300-id1.msh
  []
  [sidesets]
    type = SideSetsFromNormalsGenerator
    input =file
    normals = '0 0 -1
                0 0 1'
    fixed_normal = true
    new_boundary = 'bottom top'
  []
[]


[Physics/SolidMechanics/QuasiStatic/all]
  strain = FINITE
  add_variables = true
  save_in = 'resid_x resid_y resid_z'
[]

[BCs]
  [symmy]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0
  []
  [symmx]
    type = DirichletBC
    variable = disp_x
    boundary = bottom
    value = 0
  []
  [symmz]
    type = DirichletBC
    variable = disp_z
    boundary = bottom
    value = 0
  []
  [tdisp]
    type = FunctionDirichletBC
    variable = disp_z
    boundary = top
    function = '-1e-5*t'
  []
[]

[UserObjects]
  [prop_read]
    type = PropertyReadFile
    prop_file_name = 'input_euler.txt'
    nprop =4
    ngrain = 300
  []
[]

[Materials]
  [elasticity_tensor]
    type = ComputeElasticityTensorCP
    C_ijkl = '48.7e3 11.9e3 11.9e3 48.7e3 11.9e3 48.7e3 13.1e3 13.1e3 13.1e3'
    fill_method = symmetric9
    read_prop_user_object = prop_read
  []
  [stress]
    type = ComputeMultipleCrystalPlasticityStress
    crystal_plasticity_models = 'trial_xtalpl'
    tan_mod_type = exact
    maximum_substep_iteration = 10
  []
  [trial_xtalpl]
    type = CrystalPlasticityKalidindiUpdate
    number_slip_systems = 12
    slip_sys_file_name = input_slip_sys.txt
    ao = 2e-3
    gss_a = 2
    gss_initial = 10
    h = 100
    r = 1.4
    xm = 0.1
    t_sat = 100.
    resistance_tol = 1.0
    print_state_variable_convergence_error_messages = true
    slip_increment_tolerance = 10
  []
[]


[Preconditioning]
  [smp]
    type = SMP
    full = true
  []
[]

[Executioner]
  type = Transient
  petsc_options_iname = '-pc_type -ksp_type -ksp_gmres_restart'
  petsc_options_value = ' lu       gmres     200'
  solve_type = NEWTON
  automatic_scaling = true
  line_search = 'none'
  dtmin = 5e-12
  dtmax = 72
  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 1 # Initial time step.  In this simulation it changes.
        cutback_factor = 0.25
     growth_factor = 1.2
     optimal_iterations = 20
  [../]
  nl_abs_tol = 1e-8
  nl_rel_tol = 1e-6
  nl_max_its = 20
  end_time = 6e5 
[]

[Outputs]
  exodus = true
  interval = 70
[]
```

Also the dt stopped increasing at 16.5618 is there a way to have better convergence and increase time step.

**Comment by GiudGiud** [comment]
Hello

Do you mean the mesh has 110,000 cells? The number of dofs is tied to the number of variables. what does the simulation summary say for the number of DOFs?
Above 100k dofs it may be time to look at improving the preconditioning.

**Reply by ghost** [reply]
```
Mesh:
  Parallel Type:           replicated
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:
    Total:                 37330
    Local:                 1438
    Min/Max/Avg:           1005/1520/1244
  Elems:
    Total:                 90191
    Local:                 3093
    Min/Max/Avg:           2919/3096/3006
  Num Subdomains:          300
  Num Partitions:          30
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                111990
  Num Local DOFs:          4314
  Variables:               { "disp_x" "disp_y" "disp_z" }
  Finite Element Types:    "LAGRANGE"
  Approximation Orders:    "FIRST"
```
I have 111.9k DoF which is relatively small for a cp computation. What do you mean by improving the preconditioning

**Reply by GiudGiud** [reply]
Using a different solver package for LU or finding a different method

I wonder if you are using too many processors at this point. 4k dofs per processor is not a lot. Can you try using half as many processors to see if the performance improves?

**Reply by ghost** [reply]
Tried using mumps and 10 processors but it still takes too long.. 25min for 0.0025% strain

**Reply by sapitts** [reply]
You've identified the correct parameter to adjust for the slip tolerance, and here's the context in case that helps you decide how much more you want to increase that value:
https://github.com/idaholab/moose/blob/ad167a65116851b34db47082f452a8ba06b35cc7/modules/solid_mechanics/src/materials/crystal_plasticity/CrystalPlasticityKalidindiUpdate.C#L99

I usually balance between the slip increment tolerance, `slip_increment_tolerance`, (and when appropriate the state variable tolerance, `stol`, and the slip resistance tolerance, `resistance_tol`) and the `maximum_substep_iteration`, but I don't have a good optimization procedure. I have found that keep the `maximum_substep_iteration` at or below 5 seems to help; after more than 5 substeps the simulations seem to slow down noticeably.

**Reply by ghost** [reply]
Ok thank you. Keeping maximum_substep_iteration below 5 has helped the most.

**Comment by sapitts** [comment]
You may also like to look into the NEML2 submodule (https://mooseframework.inl.gov/modules/solid_mechanics/NEML2.html), which does have full derivatives of the model terms through libTorch. @hugary1995 can better describe the capabilities of NEML2 than I can
